core/src/main/java/org/apache/calcite/plan/Digest.java

```diff
@@ -0,0 +1,269 @@
+      List<Pair<String, Object>> items) {
+    List<Pair<String, Object>> normalized = new ArrayList<>();
+    for (Pair<String, Object> item : items) {
+      if (item.right instanceof RelNode) {
+        RelNode input = (RelNode) item.right;
+        normalized.add(Pair.of(item.left, simpleRelDigest(input)));
+      } else {
+        normalized.add(item);
+      }
+    }
+    return normalized;
+  }
+
+  /**
+   * Returns a simple string format digest.
+   *
+   * <p>There are three kinds of nodes we need to handle:
+   *
+   * <ul>
+   * <li>RelSubset: composition of class name, set id and traits;</li>
+   * <li>HepRelVertex: composition of class name and current rel id;</li>
+   * <li>Normal rel: composition of class name and id.</li>
+   * </ul>
+   *
+   * @param rel The rel
+   */
+  private static String simpleRelDigest(RelNode rel) {
+    StringBuilder digest = new StringBuilder(rel.getRelTypeName());
+    digest.append('#');
+    if (rel instanceof RelSubset) {
+      RelSubset subset = (RelSubset) rel;
+      digest.append(subset.getSetId());
+      for (RelTrait trait : subset.getTraitSet()) {
+        digest.append('.').append(trait);
+      }
+    } else if (rel instanceof HepRelVertex) {
+      digest.append(((HepRelVertex) rel).getCurrentRel().getId());
+    } else {
+      digest.append(rel.getId());
+    }
+    return digest.toString();
+  }
+
+  @Override public String toString() {
+    if (null != digest) {
+      return digest;
+    }
+    StringBuilder sb = new StringBuilder();
+    sb.append(rel.getRelTypeName());
+
+    for (RelTrait trait : rel.getTraitSet()) {
+      sb.append('.');
+      sb.append(trait.toString());
+    }
+
+    sb.append('(');
+    int j = 0;
+    for (Pair<String, Object> item : items) {
+      if (j++ > 0) {
+        sb.append(',');
+      }
+      sb.append(item.left);
+      sb.append('=');
+      sb.append(item.right);
+    }
+    sb.append(')');
+    digest = sb.toString();
+    return digest;
+  }
+
+  @Override public int compareTo(Digest other) {
+    return this.rel.getId() - other.rel.getId();
+  }
+
+  @Override public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    Digest that = (Digest) o;
+    return hashCode == that.hashCode && deepEquals(that);
+  }
+
+  /**
+   * The method is used to resolve hash conflict, in current 6000+ tests, there are about 8
+   * tests with conflict, so we do not cache the hash code items in order to
+   * reduce mem consumption.
+   */
+  private boolean deepEquals(Digest other) {
+    Object[] thisItems = collect(this.rel, this.items, true);
+    Object[] thatItems = collect(other.rel, other.items, true);
+    if (thisItems.length != thatItems.length) {
+      return false;
+    }
+    for (int i = 0; i < thisItems.length; i++) {
+      if (!Objects.equals(thisItems[i], thatItems[i])) {
```

Comment:
should Arrays.deepEquals be used if `thisItems[i]` is an array?

Reply:
I checked all the rels, they all have no array attributes in the digest, even it has, the `#equals` would output `false` which does not affect the semantics. So i choose current impl.

Reply:
for now? what about external rels?

Reply:
I said outputs false does not affect the semantics, then what's the problem there, i do not want to make the logic a mess just for a non-sure assumption. 

If you have strong opinion, please give specific cases, only assumption makes no sense to me.

Reply:
Yes, there has some "ifs" there. There are 2 assumptions here as well:
- there's no rel node using an array as a digest field, which is true for Calcite code, and probably true for code using the library although it's harder to verify
- even so, return false when 2 array contents are identical has no impact on the semantic.

For the first point, maybe a more opinionated choice could be to disallow arrays as digest item?
For the second point, I'm not sure that the assumption holds: if planner cannot identify that 2 rel nodes are identical (and so collapse them), it might cause some kind of infinite loop/plan explosion where the same node is created over and over.

Reply:
1. I don't think we should have any limitation on what should be a digest item just because of the digest code implementation
2. Did we encounter any case that has infinite loop just because of an array item there ? I need to see the case to make decision, not an assumption. I prefer to fix that when we really encounter that,  i think such case would be very rare.